internal/scan/selector.go

```diff
@@ -0,0 +1,151 @@
+	Name               string `json:"name,omitempty"`
+	LabelSelector      string `json:"labelSelector,omitempty"`
+	AnnotationSelector string `json:"annotationSelector,omitempty"`
+}
+
+// Select reduces the supplied resource node slice by only returning those
+// nodes which match this selector.
+func (g *GenericSelector) Select(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	m, err := newMetaMatcher(g)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]*yaml.RNode, 0, len(nodes))
+	for _, n := range nodes {
+		if meta, err := n.GetMeta(); err != nil {
+			return nil, err
+		} else if !m.matchesMeta(meta) {
+			continue
+		}
+
+		if matched, err := n.MatchesLabelSelector(g.LabelSelector); err != nil {
+			return nil, err
+		} else if !matched {
+			continue
+		}
+
+		if matched, err := n.MatchesAnnotationSelector(g.AnnotationSelector); err != nil {
+			return nil, err
+		} else if !matched {
+			continue
+		}
+
+		result = append(result, n)
+	}
+
+	return result, nil
+}
+
+// metaMatcher matches metadata. This is an alternative to `types.SelectorRegex` from Kustomize.
+type metaMatcher struct {
+	namespaceRegex *regexp.Regexp
+	nameRegex      *regexp.Regexp
+	groupRegex     *regexp.Regexp
+	versionRegex   *regexp.Regexp
+	kindRegex      *regexp.Regexp
+}
+
+// newMetaMatcher creates a new matcher by compiling all of the regexps.
+func newMetaMatcher(g *GenericSelector) (m *metaMatcher, err error) {
+	m = &metaMatcher{}
+
+	m.namespaceRegex, err = compileAnchored(g.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	m.nameRegex, err = compileAnchored(g.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	m.groupRegex, err = compileAnchored(g.Group)
+	if err != nil {
+		return nil, err
+	}
+
+	m.versionRegex, err = compileAnchored(g.Version)
+	if err != nil {
+		return nil, err
+	}
+
+	m.kindRegex, err = compileAnchored(g.Kind)
+	if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// matchesMeta returns true if the supplied metadata is accepted by this matcher.
+func (m *metaMatcher) matchesMeta(meta yaml.ResourceMeta) bool {
+	if m.namespaceRegex != nil && !m.namespaceRegex.MatchString(meta.Namespace) {
+		return false
+	}
+	if m.nameRegex != nil && !m.nameRegex.MatchString(meta.Name) {
+		return false
+	}
+
+	if m.groupRegex != nil || m.versionRegex != nil {
+		group, version := "", meta.APIVersion
+		if pos := strings.Index(version, "/"); pos >= 0 {
+			group, version = version[0:pos], version[pos+1:]
+		}
+
+		if m.groupRegex != nil && !m.groupRegex.MatchString(group) {
+			return false
+		}
+
+		if m.versionRegex != nil && !m.versionRegex.MatchString(version) {
+			return false
+		}
+	}
+
+	if m.kindRegex != nil && !m.kindRegex.MatchString(meta.Kind) {
+		return false
+	}
+
+	return true
+}
+
+// compileAnchored is a helper to make regular expressions match a full string.
+func compileAnchored(pattern string) (*regexp.Regexp, error) {
+	if pattern == "" {
+		return nil, nil
+	}
```

Comment:
instead of returning nil here, should we do a catchall regex ( ex, `*` or `.*` )?
that should allow us to skip the `regex != nil` checks in matchesMeta

Reply:
It just seemed like unnecessary extra use of regular expression matching in a loop. Maybe I'm just dating myself.